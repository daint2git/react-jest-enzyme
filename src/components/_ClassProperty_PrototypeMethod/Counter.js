import React from 'react'

class Counter extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      mounted: false,
      count: 0,
    }
    this.handeDecrease = this.handeDecrease.bind(this)
  }

  componentDidMount() {
    this.setState({ mounted: true })
  }

  // class property
  handeIncrease = () => this.setState(prevState => ({ count: prevState.count + 1 }))

  // prototype method
  handeDecrease() {
    this.setState(prevState => ({ count: prevState.count - 1 }))
  }

  render() {
    return (
      <>
        <button name="increment" onClick={this.handeIncrease}>
          increment
        </button>
        <button name="decrement" onClick={this.handeDecrease}>
          decrement
        </button>
        <p>{this.state.count}</p>
      </>
    )
  }
}

export default Counter
