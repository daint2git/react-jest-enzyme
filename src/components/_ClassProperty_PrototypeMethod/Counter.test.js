import Counter from './Counter'

describe('Counter', () => {
  describe('handlers', () => {
    test('should call increment method after button clicked', () => {
      const wrapper = shallow(<Counter />)
      const instance = wrapper.instance()
      const incrementMock = jest.spyOn(instance, 'handeIncrease')
      instance.forceUpdate()
      expect(incrementMock).not.toHaveBeenCalled()
      wrapper.find('button[name="increment"]').simulate('click')
      expect(incrementMock).toHaveBeenCalled()
      incrementMock.mockRestore()
    })

    test('should call increment method after button clicked (directly)', () => {
      const wrapper = shallow(<Counter />)
      const instance = wrapper.instance()
      expect(wrapper.state('count')).toBe(0)
      instance.handeIncrease()
      expect(wrapper.state('count')).toBe(1)
    })

    test('should call decrement method after button clicked', () => {
      const wrapper = shallow(<Counter />)
      const instance = wrapper.instance()
      const decrementMock = jest.spyOn(instance, 'handeDecrease')
      instance.forceUpdate()
      expect(decrementMock).not.toHaveBeenCalled()
      wrapper.find('button[name="decrement"]').simulate('click')
      expect(decrementMock).toHaveBeenCalled()
      decrementMock.mockRestore()
    })

    test('should call decrement method after button clicked (with prototype)', () => {
      const decrementMock = jest.spyOn(Counter.prototype, 'handeDecrease')
      const wrapper = shallow(<Counter />)
      expect(decrementMock).not.toHaveBeenCalled()
      wrapper.find('button[name="decrement"]').simulate('click', 'calledddddd')
      expect(decrementMock).toHaveBeenCalled()
      expect(decrementMock).toHaveBeenCalledWith('calledddddd')
      decrementMock.mockRestore()
    })
  })

  describe('should update state', () => {
    test('should update count text when increment button clicked', () => {
      const wrapper = shallow(<Counter />)
      expect(wrapper).toMatchSnapshot()
      expect(wrapper.find('p').text()).toBe('0')
      wrapper.find('button[name="increment"]').simulate('click')
      expect(wrapper.find('p').text()).toBe('1')
      expect(wrapper).toMatchSnapshot()
    })

    test('should update count in state when increment button clicked', () => {
      const wrapper = shallow(<Counter />)
      expect(wrapper).toMatchSnapshot()
      expect(wrapper.state().count).toBe(0)
      wrapper.find('button[name="increment"]').simulate('click')
      expect(wrapper.state().count).toBe(1)
      expect(wrapper).toMatchSnapshot()
    })

    test('should update mounted in state when componentDidMount', () => {
      const wrapper = shallow(<Counter />)
      expect(wrapper.state().mounted).toBe(true)
    })
  })
})
