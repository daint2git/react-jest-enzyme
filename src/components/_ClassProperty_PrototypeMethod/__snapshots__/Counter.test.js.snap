// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter should update state should update count in state when increment button clicked 1`] = `
<Fragment>
  <button
    name="increment"
    onClick={[Function]}
  >
    increment
  </button>
  <button
    name="decrement"
    onClick={[Function]}
  >
    decrement
  </button>
  <p>
    0
  </p>
</Fragment>
`;

exports[`Counter should update state should update count in state when increment button clicked 2`] = `
<Fragment>
  <button
    name="increment"
    onClick={[Function]}
  >
    increment
  </button>
  <button
    name="decrement"
    onClick={[Function]}
  >
    decrement
  </button>
  <p>
    1
  </p>
</Fragment>
`;

exports[`Counter should update state should update count text when increment button clicked 1`] = `
<Fragment>
  <button
    name="increment"
    onClick={[Function]}
  >
    increment
  </button>
  <button
    name="decrement"
    onClick={[Function]}
  >
    decrement
  </button>
  <p>
    0
  </p>
</Fragment>
`;

exports[`Counter should update state should update count text when increment button clicked 2`] = `
<Fragment>
  <button
    name="increment"
    onClick={[Function]}
  >
    increment
  </button>
  <button
    name="decrement"
    onClick={[Function]}
  >
    decrement
  </button>
  <p>
    1
  </p>
</Fragment>
`;
